packages/babel-plugin/src/utils/ast-builders.tsx

```diff
@@ -145,6 +145,109 @@ const traverseStyledBinaryExpression = (node: t.BinaryExpression, nestedVisitor:
   return node;
 };
 
+/**
+ * Handles cases like:
+ * 1. `propz.loading` in `border-color: \${(propz) => (propz.loading ? colors.N100 : colors.N200)};`
+ * Outcome: It will replace `propz.loading` with `props.loading`.
+ *
+ * 2. `prop.notValidProp` in `border-color: \${(props) => (props.notValidProp ? colors.N100 : colors.N200)};`
+ * Outcome: It will move `notValidProp` under `propsToDestructure` and replaces `props.notValidProp` with `notValidProp`.
+ *
+ * @param path MemberExpression path
+ * @param propsToDestructure Array which gets mutated with props which are not valid eg. props.notValidProp
+ */
+const handleMemberExpressionInStyledInterpolations = (
+  path: NodePath<t.MemberExpression>,
+  propsToDestructure: string[]
+) => {
+  const memberExpressionKey = path.node.object;
+
+  if (!t.isIdentifier(memberExpressionKey)) {
+    return;
+  }
+
+  const traversedUpFunctionPath = path.find((parentPath) => parentPath.isFunction());
+  const memberExpressionKeyName = memberExpressionKey.name;
+
+  const isMemberExpressionKeyNameEqualsFunctionParams =
+    t.isFunction(traversedUpFunctionPath.node) &&
+    t.isIdentifier(traversedUpFunctionPath.node.params[0]) &&
+    traversedUpFunctionPath.node.params[0].name === memberExpressionKeyName;
+
+  if (isMemberExpressionKeyNameEqualsFunctionParams) {
+    const memberExpressionValue = path.node.property;
+
+    if (!t.isIdentifier(memberExpressionValue)) {
+      return;
+    }
+
+    const memberExpressionValueName = memberExpressionValue.name;
+
+    if (isPropValid(memberExpressionValueName)) {
+      // Convert cases like propz.color to props.color
+      if (memberExpressionKeyName !== 'props') {
+        path.replaceWith(
+          t.memberExpression(t.identifier('props'), t.identifier(memberExpressionValueName))
+        );
+      }
+
+      return;
+    }
+
+    if (!propsToDestructure.includes(memberExpressionValueName)) {
+      propsToDestructure.push(memberExpressionValueName);
+    }
+
+    path.replaceWith(memberExpressionValue);
+  }
+};
+
+/**
+ * Handles cases like:
+ * 1. `isLoading` in `background-color: \${({ isLoading }) => (isLoading ? colors.N20 : colors.N40)};`
+ * Outcome: It will move `isLoading` under `propsToDestructure`.
+ *
+ * 2. `l` in `color: \${({ loading: l }) => (l ? colors.N50 : colors.N10)};`
+ * Outcome: It will move `loading` under `propsToDestructure` and replaces `l` with `loading`.
+ *
+ * @param path Identifier path
+ * @param propsToDestructure Array which gets mutated with props which extracted from destructuring
+ */
+const handleDestructuringInStyledInterpolations = (
+  path: NodePath<t.Identifier>,
+  propsToDestructure: string[]
+) => {
+  // We are not interested in parent object property like `({ loading: l }) => l`
```

Comment:
is this comment wrong?

Reply:
So in this statement, `({ loading: l }) => l` when we are checking for `Identifier`, both `: l` and `=> l` are identifier. And function is parent for both. We are not interested in modifying `: l`. We just need to modify `=> l` to `=> loading`. If we don't skip, second didn't get modify because we have modified `: l` earlier and second identifier one is nowhere to be found inside function params.
Hope this makes sense 😅

Reply:
ahhh 🤔 

so why don't we add `loading` to props to de-structure here?

alternatively now that i think about it - should we be adding destructured props anyway? currently the behaviour should be

1. am i a valid HTML element? you will be passed down to the raw HTML element
2. am i not valid? you will be destructured

do we need to rework anything to support this behaviour? 🤔